migration-guides/0.15-Migration-Guide.md

```diff
@@ -446,18 +504,169 @@ case foo of
+
+`launchAff_ :: forall a. Aff a -> Effect Unit` will _silently_ discard what the `Aff` computation returns. This behavior isn't always desirable, but the developer won't know otherwise.
+
+Thus, `launchAff_`'s type signature was changed to `Aff Unit -> Effect Unit`. By making this change, developers are forced to clarify whether they want this discarding behavior or not. It's a minor annoyance (and breaking change) for most developers but otherwise makes the resulting code much safer.
+
+**To fix**:
+- If the "discarding" behavior is desired, replace `launchAff_ do` with `launchAff_ $ void do`
+- If undesired, now you know where to fix your code
+
+### TODO - Miscellaneous changes in `purescript-css`
+
+- Add support for `calc` expressions (#140 by @nsaunders)
+- Add table selector (#141 by @plurip-software)
+- Update the box-shadow implementation (#88 by @vyorkin)
+- Dropped almost all named colors (#156 by @JordanMartinez)
+
+  These colors were originally defined in `purescript-colors` (i.e.
+  one module per schema) because each schema should be defined as its
+  own library. This change was propagated to this release.
+
+### `ParserT` got (up to) 20x performance boost (`purescript-parsing`)
+
+`ParserT` now has a more efficient representation. In addition to the performance, all parser execution is always stack-safe, even monadically, obviating the need to run parsers with `Trampoline` as the base Monad or to explicitly use `MonadRec`.
+
+Code that was parametric over the underlying Monad no longer needs to propagate a Monad constraint.
+
+Code that constructs parsers via the underlying representation will need to be updated, but otherwise the interface is unchanged and parsers should just enjoy the speed boost.
+
+**To fix**:
+- Changes only needed if you depend on `ParserT`'s underlying representation
 
 ## Breaking Changes in the `purescript-node` libraries
 
-This section has yet to be written
+### Expose the `recursive` field in `mkdir'`'s options arg (`purescript-node-fs`/`purescript-node-fs-aff`)
+
+Previously, `mkdir` did not expose all the fields in the options arg. So, trying to run something like `mkdir "foo/bar/baz" { recursive: true }` wasn't possible without writing your own FFI to `mkdir`. The `recursive` option has now been exposed.
+
+**To fix**:
+- update `mkdir' path perms` to `mkdir' path { recursive: false, mode: perms }`
+- to get just the recursive behavior, use `mkdir' path { recursive: true, mode: mkPerms all all all }`
 
 ## Breaking Changes in the `purescript-web` libraries
 
-This section has yet to be written
+### TODO - `purescript-canvas`
+
+- Support arcs that are drawn counter-clockwise (#58, #83 by @karljs and @JordanMartinez)
+- The `Transform` type now uses the field names `a`, `b`, `c`, `d`, `e` and `f`, instead of `m11`, `m12`, `m21`, `m22`, `m31` and `m32` (#86 by @artemisSystem)
+
+### Arg order for `CSSStyleDeclaration` functions changed (`purescript-cssom`)
```

Comment:
Related to my earlier comment -- I don't think we should include content for libraries outside of core, except for particularly pervasive libraries if they had a big change (ie. `aff`). If we want to add more content covering non-core libraries, that content should go in the respective organizations (for example, a `purescript-contrib` migration guide) and they can be linked from this migration guide. The migration guide for the compiler & core organization should only include compiler & core.

Reply:
Reasons why I'm against doing this "split the content across the orgs" idea:
- Why split this content into four files that disables a 'find, read, and fix" workflow? Since developers don't necessarily know whether `purescript-foo` is part of the `core` or `contrib` libs (node and web are usually obvious), they have to check two places. Won't these two issues slow down the ecosystem update? Won't that slow down packages from being added to the package set again once they've been fixed? What problem does this solve?
- Where do changelogs for `node` and `web` go since they don't have a governance repo? Is creating a repo for something this small worth it?
- If we're going to be inconsistent (e.g. "We'll only do core libraries and heavily used libraries like `aff`), what counts as 'heavily used? Most code can't be built with "just" core libraries (e.g. Halogen won't compile unless `contrib` libs are also updated, and Halogen is used by a lot of people). Why is one library included but another not?
- Making breaking changes doesn't have a corresponding "guide" (e.g. a release guide in a repo). AFAIK, even writing migration guides in 0.14 was a new thing. I'd argue for making things easier on developers for the time being and figuring out exactly how we want to do breaking changes later.

Reply:
> that disables a 'find, read, and fix" workflow? Since developers don't necessarily know whether purescript-foo is part of the core or contrib libs (node and web are usually obvious), they have to check two places.

That’s a good point. Contrib libraries especially can be confusing to find. I think this is a good reason to keep the content here.

> Why is one library included but another not?

The line I’m trying to tread is this:

1. The core team is responsible for the compiler and core libraries only
2. Some libraries, despite not technically being under our purview, arguably should be included simply because they are basically a core library (e.g. Aff), though maybe this is more an argument to move Aff into core

I’ve also been wanting to keep the signal to noise ratio high in these documents by only including “important” changes (you can see the individual library for all the details). But that’s not so relevant for an aggregated changelog.

———

All this said: I retract, and agree it’s OK to include non-core things in these change summaries. Thanks for helping me see some points I wasn’t considering @JordanMartinez

Reply:
> Some libraries, despite not technically being under our purview, arguably should be included simply because they are basically a core library (e.g. Aff), though maybe this is more an argument to move Aff into core

Doesn't `aff` exist in `contrib` because some backends don't need it whereas the JS backend does? Or put differently, if PureScript formalized a runtime for backends to implement, `aff` could be in core, right?